### ISSUE TITLE
Dropout rescaling

### ISSUE THREAD

**Issue by benanne** [benanne/Lasagne#10]
`nntools.layers.DropoutLayer` currently implements dropout in a nonstandard way. During training (`determinstic=False`), a Bernoulli dropout mask is applied to the input (with retention probability `1 - p`) and the input is then scaled by dividing it by `1 - p`. At test time (`deterministic=True`), the layer does nothing and just passes through its input.

This is different from the typical approach, where the mask is applied during training and rescaling by multiplying with `1 - p` happens at test time.

I decided to do the rescaling during training, instead of at test time. I think this is cleaner because then dropout is something that only affects training. At test time nothing changes, regardless of whether dropout is used or not.

Also, the weight initialization does not need to be changed depending on whether dropout is used.

That said, although I think this is the cleanest way to implement dropout, it might be confusing for some users. So maybe we should do the rescaling at test time instead. Or maybe this should be an option?

I also included an option to disable the rescaling altogether, which is useful for implementing denoising autoencoders (no rescaling is typically used in that context).

#9 is also related to this issue.

**Comment by f0k** [comment]
> Not sure I follow you about having to change the learning rate for the next layer.

I think the comment in the pylearn2 issue (and thus my comment above) was wrong. We scale the input symbolically, i.e., our forward pass for dropout -> dense ends up being `np.dot(2 * input * mask, weights) + bias` instead of `np.dot(input, weights) + bias`. When computing the gradient wrt the weights, Theano will thus include the correct scaling for the gradient anyway and there is no need to adapt the learning rate(s).
